packages/react-components/src/gallery/dominantSpeaker.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { VideoGalleryRemoteParticipant } from '../types';
+
+/**
+ * Calculates the participants that should be renderd basis on the list of dominant
+ * speakers and currently renderd participants in a call.
+ * @param participants - Array containing all participants of a call.
+ * {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param dominantSpeakers - An array containing the userId of dominant speakers
+ * in a call in the order of their dominance. 0th index is the most dominant, 1st is the second most etc
+ * @param visibleParticipants - Array containing currently rendered (visible)
+ * participants in the call. {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ * @param maxTiles - Maximum number of tiles to calculate.
+ * @returns VideoGalleryRemoteParticipant[] {@link @azure/communication-react#VideoGalleryRemoteParticipant}
+ */
+export const smartDominantSpeakerParticipants = (
+  participants: VideoGalleryRemoteParticipant[] = [],
+  dominantSpeakers: Array<string> = [],
+  visibleParticipants: VideoGalleryRemoteParticipant[] = [],
+  maxTiles = 4
+): VideoGalleryRemoteParticipant[] => {
+  if (!participants) return [];
+
+  // Don't apply any logic if total number of video streams is less than Max video streams.
+  if (participants.length <= maxTiles) return participants;
+
+  let currentParticipants = visibleParticipants.slice();
+  // Initialize `currentParticipants` if it is empty.
+  if (!currentParticipants.length) {
+    currentParticipants = participants.slice(0, maxTiles);
```

Comment:
Is there a reason why we do `participants.slice(0,4)` but just `visibleParticipants.slice()` (without the 0,4)?

Reply:
Is it OK that visible participants might be less than `maxTiles`?

Reply:
Shallow copying visibleParticipants using slice to avoid mutating it as its a function parameter.

Reply:
And yes, it's ok for visibleParticipants to be less than maxTiles.

Reply:
Oh right currentParticipants are the array of currently visible participants, which might be the array _or_ (we assume it to be) the maxTiles number of currentParticipants.

Can we avoid this second slice and just allow currentParticipants to be an empty array if visibleParticipants.length === 0? (might be wrong) but if its an empty array, looks like it will first push the dominant speakers onto the array in order, then participants.

Reply:
Yes, it can be removed and I have removed it :)
Thanks for pointing it out.